server/dbutil/dbutil.go

```diff
@@ -0,0 +1,161 @@
+		createFilePairs = `CREATE TABLE IF NOT EXISTS file_pairs (
+			id INTEGER, name_a TEXT, name_b TEXT, hash_a TEXT, hash_b TEXT,
+			content_a TEXT, content_b TEXT, diff TEXT,experiment_id INTEGER,
+			PRIMARY KEY (id),
+			FOREIGN KEY(experiment_id) REFERENCES experiments(id))`
+		createAssignments = `CREATE TABLE IF NOT EXISTS assignments (
+			user_id INTEGER, pair_id INTEGER, experiment_id INTEGER,
+			answer INTEGER, duration INTEGER,
+			PRIMARY KEY (user_id, pair_id),
+			FOREIGN KEY (user_id) REFERENCES users(id),
+			FOREIGN KEY (pair_id) REFERENCES file_pairs(id),
+			FOREIGN KEY (experiment_id) REFERENCES experiments(id))`
+	)
+
+	if _, err := db.Exec(createUsers); err != nil {
+		return err
+	}
+
+	if _, err := db.Exec(createExperiments); err != nil {
+		return err
+	}
+
+	if _, err := db.Exec(createFilePairs); err != nil {
+		return err
+	}
+
+	if _, err := db.Exec(createAssignments); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Initialize populates the DB with default values. It is safe to call on a
+// DB that is already initialized
+func Initialize(db *sql.DB) error {
+	_, err := db.Exec(`
+		INSERT OR IGNORE INTO experiments (id, name, description)
+		VALUES ($1, 'default', 'Default experiment')`,
+		defaultExperimentID)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Options for the ImportFiles method. Logger is optional, if it is not provided
+// the default stderr will be used.
+type Options struct {
+	Logger *log.Logger
+}
+
+// ImportFiles imports pairs of files from the origin to the destination DB.
+// It copies the contents and processes the needed data (md5 hash, diff)
+func ImportFiles(originDB, destDB *sql.DB, opts Options) (success, failures int64, e error) {
+	logger := log.New(os.Stderr, "", log.LstdFlags) // Default log to stderr
+
+	if opts.Logger != nil {
+		logger = opts.Logger
+	}
+
+	rows, err := originDB.Query("SELECT * FROM files;")
+	if err != nil {
+		return 0, 0, err
+	}
+
+	tx, err := destDB.Begin()
+
+	insert, err := tx.Prepare(`INSERT INTO file_pairs
+		(name_a, name_b, hash_a, hash_b, content_a, content_b, diff, experiment_id)
+		VALUES ($1, $2, $3, $4, $5, $6, $7, $8);`)
+
+	if err != nil {
+		return 0, 0, err
+	}
+
+	for rows.Next() {
+		var nameA, nameB, contentA, contentB, diffText string
+		if err := rows.Scan(&nameA, &nameB, &contentA, &contentB); err != nil {
+			return success, failures, err
+		}
+
+		diffText, err := diff(nameA, nameB, contentA, contentB)
+		if err != nil {
+			logger.Printf(
+				"Failed to create diff for files:\n - %q\n - %q\nerror: %v\n",
+				nameA, nameB, err)
+			failures++
+			continue
+		}
+
+		res, err := insert.Exec(nameA, nameB,
+			md5hash(contentA), md5hash(contentB),
+			contentA, contentB,
+			diffText,
+			defaultExperimentID)
+
+		if err != nil {
+			logger.Println(err)
+			failures++
+			continue
+		}
+
+		rowsAffected, _ := res.RowsAffected()
+		success += rowsAffected
+	}
+
+	if err := tx.Commit(); err != nil {
```

Comment:
do you want to add `rollback` somewhere? I'm not sure about sqlite, but for other dbs it's better to rollback. :)

Reply:
It actually looks like Commit or Rollback should be the last call, if I understand this correctly Commit will close the transaction even if it fails:
https://golang.org/src/database/sql/sql.go?s=49331:49359#L1761

Reply:
You do `return success, failures, err` above. So neither Commit neither Rollback will be executed. And we will end up with hanging transaction.

Reply:
Yes, missed that return inside the loop, thank you.